### DISCUSSION TITLE
useIsFetching() returns 0 when isFetching from query is true

### DISCUSSION THREAD

**Discussion by amie-wilt** [TanStack/query#2061] [Q&A]
**Describe the bug**
Both `useIsFetching()` and `queryClient.isFetching()` are `0` when `isFetching` from a `useQuery` call is `true`

**To Reproduce**

**Expected behavior**
I expect `useIsFetching()` to return `1`

**Screenshots**

Console log statement for `useIsFetching()` call:
![Screen Shot 2021-03-31 at 9 49 52 AM](https://user-images.githubusercontent.com/8226176/113155283-dbcc7200-9206-11eb-99f1-28108fff1ec8.png)

Console log statement for `<RecentPublication />`:
![Screen Shot 2021-03-31 at 9 50 17 AM](https://user-images.githubusercontent.com/8226176/113155293-e0912600-9206-11eb-876f-59fadfbd5811.png)

`useFetchCommentary`:
![Screen Shot 2021-03-31 at 9 50 25 AM](https://user-images.githubusercontent.com/8226176/113155329-ea1a8e00-9206-11eb-9e99-8878798bbae5.png)

Resultings logs:
<img width="380" alt="Screen Shot 2021-03-31 at 9 51 14 AM" src="https://user-images.githubusercontent.com/8226176/113155396-f9014080-9206-11eb-9b72-57916d483be6.png">

**Desktop (please complete the following information):**
 - Browser: [e.g. chrome, safari]
 - Version: [e.g. 22]

**Smartphone (please complete the following information):**
 - Browser: [e.g. stock browser, safari]
 - Version: [e.g. 22]

**Additional context**

**Comment by TkDodo** [comment]
1) you are logging the `recent pub isFetching` during render, and _not_ in a useEffect. That's why it prints _before_ the `isFetching after mount`. if you log both in an effect, you will likely see that you first get `after mount 0` followed by `recent true`

2) the `isFetching after mount` prints false because when the component mounts, it is false. You are depending on the result of `useIsFetching`, but your effect dependencies don't represent that. The eslint rule `"react-hooks/exhaustive-deps` should flag that.

**Comment by amie-wilt** [comment]
@TkDodo Thanks for the response, I'm sure I'm missing something here so pardon my misunderstanding on this.

1. I moved `recent pub isFetching` to a `useEffect` with no deps for an "on mount" action but am still seeing it print first, which is what I would expect?:

![Screen Shot 2021-03-31 at 10 36 11 AM](https://user-images.githubusercontent.com/8226176/113162428-4d0f2380-920d-11eb-9c47-dd7a10882f40.png)
<img width="217" alt="Screen Shot 2021-03-31 at 10 35 53 AM" src="https://user-images.githubusercontent.com/8226176/113162476-54cec800-920d-11eb-9afe-52a7df8bc7f5.png">

2. Why would `isFetching after mount` be false when the component mounts? Again, I think I'm probably missing something here but wouldn't the component be considered mounted after it's children mounted? Therefore `<RecentPublication />` mounts and fires `recent pub isFetching` first and then `isFetching after mount` second in its parent?

**Comment by TkDodo** [comment]
no worries, I think I closed prematurely and should've moved this to a discussion. apologies.

if `RecentPublication` is a child of `Dashboard`, and they both do `useEffects`, I'm relatively certain that the effect from the parent should run _before_ the effect of the child. can you reproduce this in a codesandbox? Anyhow, if you want to log `isFetching`, I would set the dependency array accordingly:

```
const isFetching = useIsFetching()
useEffect(() => {
  console.log('isFetching count', isFetching)
}, [isFetching]
```

**Reply by amie-wilt** [reply]
I did a quick sandbox for child/parent mounting order: https://codesandbox.io/s/frosty-ellis-xs02b?file=/src/Child.js
Also, this is in a react native app, not sure it matters.

I'm trying to show a "global" loader for a dashboard view with cards. Instead of calling `useQueries` and passing down the results to children (cards), I'm `useQuery`ing individually in each dashboard card (because they're also used elsewhere) with the idea I'd be able to use `useIsFetching` to determine the fetching state in the parent. 

So in Dashboard I'd have:

```
const [loading, setLoading] = useState(true);

useEffect(() => {
if (isFetching === 0 && loading) setLoading(false)
}, [isFetching])
```

But because `isFetching` is `0` immediately, `loading` is set to `false` immediately and my loading state is moot.

**Comment by TkDodo** [comment]
yes, so in the parent component, on the first render, `isFetching` will be `0`. Then, once a query really fetches, this will go to `1`.

I don't think you need _state_ for the `loading` indicator. You can just do:

```
const isFetching = useIsFetching()
const loading = isFetching > 0
```

loading is a _derived_ state from `isFetching`

**Reply by amie-wilt** [reply]
Fair enough... I started overcomplicating this when I implemented `refreshing` with react native's `<RefreshControl />` (don't want to show loader when `refreshing`) but I can obviously use the `refreshing` flag so that `const loading = isFetching > 0 && !refreshing`. 

Thanks for your time on this!
